### DISCUSSION TITLE
Options for radios without built-in PARA wireless trainer

### DISCUSSION THREAD

**Discussion by ysoldak** [ysoldak/HeadTracker#31] [Ideas]
While the Head Tracker tested and works nicely with recent FrSky radios that have PARA wireless trainer built-in, there have been [several cases](https://github.com/ysoldak/HeadTracker/issues/30) recently people can't connect Head Tracker to their modified Radiomaster TX16S.
The modification involves [installing a bluetooth module inside the radio](https://headtracker.gitbook.io/head-tracker/getting-started/wiring/option-1-bluetooth/adding-a-frsky-bt-module).

Apparently, there are different bluetooth modules on the market that look almost the same and only one of them ([most expensive](https://www.horusrc.com/en/frsky-horus-x10-para-wireless-module.html), of course) shall actually work, w/o need for flashing any firmware on it.

I can not verify this claim myself, due to lack of hardware, but according to one gentleman in [discussion of the issue in Cliff's Head Tracker](https://github.com/dlktdr/HeadTracker/issues/22), the module with "PARA" in name and price tag of ~$30 actually works. This is what [he has to say](https://github.com/dlktdr/HeadTracker/issues/22#issuecomment-808720210):
> BTW I can confirm not just T16 but also T16s works with "[FrSky ACCESS PARA Wireless Module](https://www.horusrc.com/en/frsky-horus-x10-para-wireless-module.html)"

---

Take a look at the [WRONG module](https://www.horusrc.com/en/frsky-transmitter-horus-x12s-bluetooth-module.html), that shall not work with this Head Tracker, so DO NOT buy it!

Worth to mention, [Cliff's Head Tracker](https://github.com/dlktdr/HeadTracker) does work with $10 bluetooth module, so if you already soldered the _wrong_ module to your radio, your best bet would be use that head tracker firmware.

I have no plans to implement support for $10 bluetooth module in this project, at least not myself.
But feel free to contributes code that works with that module, make a PR and we take it from there.

---

The options you have with radios w/o built-in PARA wireless trainer:
- Use `PPM`, [wired connection](https://github.com/ysoldak/HeadTracker/tree/master#wire);
- Use `Arduino Nano 33 BLE` on the radio side as a helper module (from [Cliff's project](https://github.com/dlktdr/HeadTracker)), HT will connect to it wirelessly (same protocol) and helper module will connect to your radio via SBUS or PPM;
- Use `proper PARA module` installed in your radio;
- Use an `ESP32 board` in your radio and flash [BTWifiModule](https://github.com/dlktdr/BTWifiModule) firmware on it.

**Comment by Wolle2602** [comment]
Hi ysoldak,
recently I have bought your Walksnail 3D print with the Seeed Studio XIAO BLE Sense board and flashed release v2.4.3 to it. In my Radio (TX16S) I have the diy ESP32 board installed (BTWifiModule). 
It works, but I have periodical connection losses. On another goggle I have an Arduino Nano 33 BLE (Cliff's code) which works fine with the ESP32...
Do you know where the problem is?

**Reply by ysoldak** [reply]
"Periodical connection losses" -- does connection restore itself or you need to re-connect when lost?

I have no first-hand experience with `BTWifiModule`, so can not reproduce and can only speculate.

The connection stability depends on EM noise in vicinity of BTWifiModule inside your radio. Also, the relative position and distance between BT devices. You have `Arduino Nano 33 BLE` mounted differently, I guess, probably on a side of goggles?

I myself did not experience problems with connectivity with case mounted on top or goggles but my radio supports PARA wireless trainer directly, so I do not use `BTWifiModule`.

Try reposition the HT case (I know, it is designed to be on top of goggles, but for test sake). Next test would be to try and flash different firmware, I believe Cliff's HT supports XIAO BLE too.

**Reply by Wolle2602** [reply]
Hi ysoldak,
it reconnects instantly - but enough to turn servos to neutral position.
The HT is still not mounted on the goggle, so this is not the issue. But maybe the EM noise could be the problem.
Around and on my desk there are about 10 devices having Wifi and BT - but strange that the Arduino is not having connection problems.
I will test on the airfield and then maybe try other firmware.

(Btw. I remember Cliff had problems with his early firmware and the Frsky BT module (not PARA) and the issue was something about Package rates - just to think about it?! You are the Expert!)

**Reply by ysoldak** [reply]
Arduino Nano 33 BLE is a different board and may have antenna positioned better, etc.

If HT on XIAO BLE board reconnects instantly, it means the firmware does not reboot or hang, so must be something with Bluetooth connection. 

Not familiar with Frsky BT module (not PARA), probably not relevant here...

Have you had a chance to try on the open, away from other devices?

**Reply by Wolle2602** [reply]
Hi ysoldak,
I tried somewhere without frequency noise, didn't change a thing.
I also bought a 2nd Xiao board to find out if it is damaged - it is not...
Next thing I can try is wiring a PARA module to my Radio (have one somewhere...).
(TX16S MKII mainboard is not made for that)

Want your Module to work, but still don't have a clue where the Problem might be.

**Reply by ysoldak** [reply]
Sorry can't help you here. Would need to reproduce first, but it works stable for me...

Some random idea: if you detach screen, button and all and just power XIAO BLE (with HT firmware flashed) via USB-C cable. Connect to it from your radio and see if link is stable?  What I'm thinking is the battery, or display or something else around XIAO can disturb bluetooth waves / block antenna. Experiment with distance from your radio...

Anyway, good to have the information here, so others can see there may be a link issue between PARA module installed in TX16S and XIAO.

**Reply by Wolle2602** [reply]
Last reply, because it works now! I have installed the PARA BT module in my Radio now...

I tried all your ideas but at the end my only thought is that maybe BTWifiModule and your XIAO board/firmware are not compatible...

**Reply by ysoldak** [reply]
Great! Happy head tracking!
Yeah, I probably misunderstood you from start. It is known to work with PARA BT module and that ESP32 board is just an experiment (or at least it was not long time ago).
